### PR TITLE
[Gekidou DB]: Adds threads in team database table and handlers

### PR DESCRIPTION
**Pull request by koox00**
#### Summary

DM/GM channels have no team ID. This makes it troublesome to paginate
threads in a team. The issue is that whenever a DM/GM thread is fetched
from pagination it will be added in all teams in that server,
potentially creating gaps in between threads for those teams.

This PR inserts a new table in the DB ThreadsInTeam which will hold
references of threads loaded in which server.
Thread lists then would have to rely on that table to show threads.

#### DB Schema

Screenshot 2022-03-24 at 4 58 27 PM

#### Checklist

- [x] Added or updated unit tests (required for all new features)
- [ ] Has UI changes
- [ ] Includes text changes and localization file updates
- [ ] Have tested against the 5 core themes to ensure consistency between them.

#### Device Information
This PR was tested on: 

#### Release Note

```release-note
NONE
```

### CODE REVIEW

File: app/database/models/server/index.ts [hunk 1]
```diff
@@ -26,5 +26,6 @@ export {default as TeamModel} from './team';
 export {default as TeamSearchHistoryModel} from './team_search_history';
 export {default as TermsOfServiceModel} from './terms_of_service';
 export {default as ThreadModel} from './thread';
+export {default as ThreadInTeamModel} from './team_thread';
```
Comment (enahum) [inline app/database/models/server/index.ts:29]: Can we rename the file to `threads_in_team` to match the model name, or change the model name, you decide

File: app/database/models/server/team.ts [hunk 2]
```diff
@@ -102,4 +107,21 @@ export default class TeamModel extends Model {
 
     /** teamSearchHistories : All the searches performed on this team */
     @children(TEAM_SEARCH_HISTORY) teamSearchHistories!: TeamSearchHistoryModel[];
+
+    /** threads : All threads belonging to a team */
+    @lazy threads = this.collections.get(THREAD).query(
```
Comment (enahum) [inline app/database/models/server/team.ts:112]: Is this missing a type?

File: app/database/operator/server_data_operator/handlers/thread_in_team.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        if (!Object.keys(threadsMap).length) {
+            return [];
+        }
+
+        const create: ThreadInTeam[] = [];
+        const teamIds = Object.keys(threadsMap);
+        for await (const teamId of teamIds) {
+            const chunks = (await this.database.get(THREADS_IN_TEAM).query(
+                Q.where('team_id', teamId),
+            ).fetch()) as ThreadInTeamModel[];
```
Comment (enahum) [inline app/database/operator/server_data_operator/handlers/thread_in_team.ts:30]:
```suggestion
            const chunks = (await this.database.get<ThreadInTeamModel>(THREADS_IN_TEAM).query(
                Q.where('team_id', teamId),
            ).fetch());
```

**Review by enahum — CHANGES_REQUESTED** [review]

File: app/database/operator/server_data_operator/handlers/thread_in_team.ts (hunk 3 above)
Comment (koox00) [reply app/database/operator/server_data_operator/handlers/thread_in_team.ts:30]: for some dark reason this does not work..

File: app/database/operator/server_data_operator/transformers/thread.ts [hunk 4]
```diff
@@ -1,6 +1,7 @@
 // Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
 // See LICENSE.txt for license information.
 
+import ThreadInTeamModel from '@app/database/models/server/thread_in_team';
 import {MM_TABLES} from '@constants/database';
 import {prepareBaseRecord} from '@database/operator/server_data_operator/transformers/index';
```
Comment (enahum) [inline app/database/operator/server_data_operator/transformers/thread.ts:6]:
```suggestion
import {MM_TABLES} from '@constants/database';
import {prepareBaseRecord} from '@database/operator/server_data_operator/transformers/index';

import type ThreadInTeamModel from '@typings/database/models/server/thread_in_team';
```

File: types/database/models/servers/team_thread.d.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {Relation} from '@nozbe/watermelondb';
```
Comment (enahum) [inline types/database/models/servers/team_thread.d.ts:4]: rename this file

File: app/database/operator/server_data_operator/handlers/thread_in_team.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {Q, Database} from '@nozbe/watermelondb';
+
+import {Database as DBConstants} from '@constants';
+import {transformThreadInTeamRecord} from '@database/operator/server_data_operator/transformers/thread';
+import {getRawRecordPairs} from '@database/operator/utils/general';
+import {HandleThreadInTeamArgs} from '@typings/database/database';
```
Comment (enahum) [inline app/database/operator/server_data_operator/handlers/thread_in_team.ts:9]:
```suggestion
import type {HandleThreadInTeamArgs} from '@typings/database/database';
```

File: app/database/operator/server_data_operator/handlers/thread_in_team.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+import type ThreadInTeamModel from '@typings/database/models/servers/team_thread';
```
Comment (enahum) [inline app/database/operator/server_data_operator/handlers/thread_in_team.ts:11]: rename this file

File: app/database/operator/server_data_operator/handlers/thread_in_team.ts [hunk 8]
```diff
@@ -0,0 +1,62 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {Q, Database} from '@nozbe/watermelondb';
+
+import {Database as DBConstants} from '@constants';
```
Comment (enahum) [inline app/database/operator/server_data_operator/handlers/thread_in_team.ts:6]: you could do `import {MM_TABLES} from '@constants/database';`

**Review by enahum — CHANGES_REQUESTED** [review]

**Review by enahum — APPROVED** [review]

File: app/database/models/server/thread_in_team.ts [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+import type ThreadModel from '@typings/database/models/servers/thread';
+
+const {TEAM, THREAD, THREADS_IN_TEAM} = MM_TABLES.SERVER;
+
+/**
+ * ThreadInTeam model helps us to combine adjacent threads together without leaving
+ * gaps in between for an efficient user reading experience for threads.
+ */
+export default class ThreadInTeamModel extends Model {
+    /** table (name) : PostsInThread */
```
Comment (avinashlng1080) [inline app/database/models/server/thread_in_team.ts:20]:
```suggestion
    /** table (name) : ThreadsInTeam */
```

**Review by avinashlng1080 — APPROVED** [review]
LGTM.  One small nit

**Review by anurag6713 — APPROVED** [review]

**Pull request merged**